chain/chain/src/resharding/trie_state_resharder.rs

```diff
@@ -0,0 +1,260 @@
+        tracing::info!(
+            target: "resharding",
+            ?left_state_root,
+            ?right_state_root,
+            ?event.left_child_shard,
+            ?event.right_child_shard,
+            "TrieStateResharding: child state roots"
+        );
+
+        let mut status = TrieStateReshardingStatus::new(
+            event.parent_shard,
+            TrieStateReshardingChildStatus::new(event.left_child_shard, left_state_root),
+            TrieStateReshardingChildStatus::new(event.right_child_shard, right_state_root),
+        )
+        .with_metrics();
+        self.resharding_blocking_impl(&mut status)
+    }
+
+    /// Resume an interrupted resharding operation.
+    pub fn resume(&self, shard_uid: ShardUId) -> Result<(), Error> {
+        let Some(status) = self.load_status()? else {
+            tracing::info!(target: "resharding", "Resharding status not found, nothing to resume.");
+            return Ok(());
+        };
+
+        if status.shard_uid != shard_uid {
+            tracing::error!(
+                target: "resharding", status_shard_uid=?status.shard_uid, ?shard_uid,
+                "Resharding status shard UID does not match the provided shard UID.");
+            return Err(Error::ReshardingError(format!(
+                "Resharding status shard UID {} does not match the provided shard UID {}.",
+                status.shard_uid, shard_uid
+            )));
+        }
+        let mut status = status.with_metrics();
+        self.resharding_blocking_impl(&mut status)
+    }
+
+    fn resharding_blocking_impl(
+        &self,
+        status: &mut TrieStateReshardingStatus,
+    ) -> Result<(), Error> {
+        while !status.done() && !self.handle.is_cancelled() {
+            self.process_batch_and_update_status(status)?;
+        }
+
+        Ok(())
+    }
+}
+
+impl Debug for TrieStateResharder {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("TrieStateResharder").field("handle", &self.handle).finish()
+    }
+}
+
+/// Metrics for tracking store column update during resharding.
+struct TrieStateResharderMetrics {
+    processed_batches: IntGauge,
+}
+
+impl TrieStateResharderMetrics {
+    pub fn new(shard_uid: &ShardUId) -> Self {
+        let processed_batches = trie_state_metrics::STATE_COL_RESHARDING_PROCESSED_BATCHES
+            .with_label_values(&[&shard_uid.to_string()]);
+        Self { processed_batches }
+    }
+
+    pub fn inc_processed_batches(&self) {
+        self.processed_batches.inc();
+    }
+}
+
+fn next_batch(
+    tries: ShardTries,
+    shard_uid: ShardUId,
+    state_root: CryptoHash,
+    seek_key: Vec<u8>,
+    batch_size: usize,
+    store_update: &mut TrieStoreUpdateAdapter,
+) -> Result<Option<Vec<u8>>, StorageError> {
+    let trie = tries.get_trie_for_shard(shard_uid, state_root).recording_reads_new_recorder();
+    let locked = trie.lock_for_iter();
+    let mut iter = locked.iter()?;
+    iter.seek(seek_key, true)?;
```

Comment:
nit: Can you change this boolean to an enum with nicely named variants? I'm guessing it's whether the iter is inclusive or exclusive but it's hard to say from here.